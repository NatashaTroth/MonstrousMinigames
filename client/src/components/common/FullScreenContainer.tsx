import * as React from 'react'

import { StyledFullScreenContainer } from './FullScreenContainer.sc'

const FullScreenContainer: React.FunctionComponent = ({ children }) => (
    <StyledFullScreenContainer>{children}</StyledFullScreenContainer>
)

export default FullScreenContainer
