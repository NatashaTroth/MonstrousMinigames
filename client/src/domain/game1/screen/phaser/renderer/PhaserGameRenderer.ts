//TODO can be used by all phaser games!!

import Phaser from 'phaser';

import { designDevelopment } from '../../../../../utils/constants';
import { depthDictionary } from '../../../../../utils/depthDictionary';
import { getRandomInt } from '../../../../../utils/getRandomInt';
import MainScene from '../../components/MainScene';
import * as colors from '../colors';
import { gameLoadedWaitingMessages, gameLoadingMessages } from '../gameLoadingMessages';
import { countdownTextStyleProperties, loadingTextStyleProperties } from '../textStyleProperties';

/**
 * this is an incomplete GameRenderer adapter which contains all the phaser logic. This class might only be tested via
 * integration tests. That's why we want to keep this class as small as possible.
 */
export class PhaserGameRenderer {
    sceneWidth: number;
    sceneHeight: number;
    progressBoxWidth: number;
    progressBoxHeight: number;
    countdownText?: Phaser.GameObjects.Text;
    progressBox?: Phaser.GameObjects.Graphics;
    progressBar?: Phaser.GameObjects.Graphics;
    loadingText?: Phaser.GameObjects.Text;
    percentText?: Phaser.GameObjects.Text;
    assetText?: Phaser.GameObjects.Text; //only for local dev -> to see which assets take long to load

    constructor(private scene: MainScene) {
        this.scene = scene;
        this.sceneWidth = this.scene.cameras.main.width;
        this.sceneHeight = this.scene.cameras.main.height;
        this.progressBoxWidth = 320;
        this.progressBoxHeight = 50;
    }

    renderCountdown(text: string) {
        const fixedWidth = 800;
        const fixedHeight = 200;
        const x = this.scene.windowWidth / 2 - fixedWidth / 2;
        const y = this.scene.windowHeight / 2 - fixedHeight / 2;

        if (this.countdownText) {
            this.countdownText.setText(text);
        } else {
            this.countdownText = this.scene.make.text({
                x,
                y,
                text,
                style: {
                    ...countdownTextStyleProperties,
                    fontSize: `${fixedHeight}px`,
                    fixedWidth,
                    fixedHeight,
                },
                add: true,
            });
            this.countdownText.scrollFactorX = 0;
            this.countdownText.setDepth(depthDictionary.countdown);
        }
    }

    renderLoadingScreen() {
        //progress bar: https://gamedevacademy.org/creating-a-preloading-screen-in-phaser-3/?a=13#Loading_Our_Assets

        //loading bar
        this.progressBar = this.scene.add.graphics();
        this.progressBox = this.scene.add.graphics();
        this.progressBox.fillStyle(0xa7bdb1);
        this.progressBox.setDepth(depthDictionary.progressBox);

        const screenCenterWidth = this.scene.cameras.main.worldView.x + this.scene.cameras.main.width / 2;
        const screenCenterHeight = this.scene.cameras.main.worldView.y + this.scene.cameras.main.height / 2;
        const progressBoxXPos = this.sceneWidth / 2 - this.progressBoxWidth / 2;
        const progressBoxYPos = this.sceneHeight / 2 - this.progressBoxHeight / 2;
        this.progressBox.fillRect(progressBoxXPos, progressBoxYPos, this.progressBoxWidth, this.progressBoxHeight);

        this.loadingText = this.scene.make.text({
            x: screenCenterWidth,
            y: screenCenterHeight - this.progressBoxHeight,
            text: `${gameLoadingMessages[getRandomInt(0, gameLoadingMessages.length)]}...`,
            style: {
                ...loadingTextStyleProperties,
                fontSize: `${20}px`,
            },
        });
        this.loadingText.setOrigin(0.5);

        //loading percentage
        this.percentText = this.scene.make.text({
            x: screenCenterWidth,
            y: screenCenterHeight,
            text: '0%',
            style: {
                ...loadingTextStyleProperties,
                fontSize: `${18}px`,
                color: colors.darkTreeGreen,
                fontStyle: 'bold',
            },
        });
        this.percentText.setOrigin(0.5);
        this.percentText.setDepth(depthDictionary.percentText);

        if (designDevelopment) {
            //asset text
            this.assetText = this.scene.make.text({
                x: this.sceneWidth / 2,
                y: this.sceneHeight / 2 + 50,
                text: '',
                style: {
                    ...loadingTextStyleProperties,
                },
            });
            this.assetText.setOrigin(0.5, 0.5);
        }
    }

    updateLoadingScreenPercent(value: number) {
        this.percentText?.setText(`${Math.round(value * 100)}%`);

        this.progressBar?.clear();
        this.progressBar?.fillStyle(0xd2a44f, 1);
        this.progressBar?.fillRect(
            this.sceneWidth / 2 - this.progressBoxWidth / 2 + 10,
            this.sceneHeight / 2 - this.progressBoxHeight / 2 + 10,
            300 * value,
            30
        );
        this.progressBar?.setDepth(depthDictionary.progressBar);
    }

    updateLoadingScreenFinishedPreloading() {
        this.loadingText?.setText(`${gameLoadedWaitingMessages[getRandomInt(0, gameLoadedWaitingMessages.length)]}...`);
    }

    //only for local development
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    fileProgressUpdate(file: any) {
        this.assetText?.setText(`Loading asset: ${file.src}`);
    }

    destroyLoadingScreen() {
        this.loadingText?.destroy();
        this.percentText?.destroy();
        this.progressBox?.destroy();
        this.progressBar?.destroy();
        if (designDevelopment) this.assetText?.destroy();
    }

    destroyCountdown() {
        this.countdownText?.destroy();
    }
}
