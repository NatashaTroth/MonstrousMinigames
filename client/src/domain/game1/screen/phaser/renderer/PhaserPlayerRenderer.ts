import Phaser from 'phaser';

import { depthDictionary } from '../../../../../utils/depthDictionary';
import { fireworkFlares } from '../../components/GameAssets';
import MainScene from '../../components/MainScene';
import * as colors from '../colors';
import { Character, CharacterAnimation } from '../gameInterfaces';
import { CharacterAnimationFrames } from '../gameInterfaces/Character';
import { Coordinates } from '../gameTypes';
import { sharedTextStyleProperties } from '../textStyleProperties';

/**
 * this is an incomplete PlayerRenderer adapter which contains all the phaser logic. This class might only be tested via
 * integration tests. That's why we want to keep this class as small as possible.
 */

interface RendererObstacle {
    phaserInstance: Phaser.Types.Physics.Arcade.SpriteWithDynamicBody;
}
export class PhaserPlayerRenderer {
    private player?: Phaser.Types.Physics.Arcade.SpriteWithDynamicBody;
    private chaser?: Phaser.Types.Physics.Arcade.SpriteWithDynamicBody;
    private wind?: Phaser.Types.Physics.Arcade.SpriteWithDynamicBody;
    private obstacles: RendererObstacle[];
    private skippedObstacles: RendererObstacle[];
    private playerAttention?: Phaser.Types.Physics.Arcade.SpriteWithDynamicBody;
    private playerWarning?: Phaser.Types.Physics.Arcade.SpriteWithDynamicBody;
    private particles: Phaser.GameObjects.Particles.ParticleEmitterManager[];
    private playerNameBg?: Phaser.GameObjects.Rectangle;
    private playerName?: Phaser.GameObjects.Text;
    private caveBehind?: Phaser.Types.Physics.Arcade.SpriteWithDynamicBody;
    private caveInFront?: Phaser.Types.Physics.Arcade.SpriteWithDynamicBody;
    private backgroundElements?: Phaser.GameObjects.Image[];
    //private backgroundLane?: Phaser.GameObjects.TileSprite[]; // for tile parallax

    constructor(
        private scene: MainScene,
        private numberPlayers: number,
        private laneHeightsPerNumberPlayers: number[]
    ) {
        this.obstacles = [];
        // Natasha bitte l√∂schen wenn das nicht mehr gebraucht wird
        this.skippedObstacles = [];
        this.particles = [];
        this.backgroundElements = []; //TODO change

        //when <= 2 lanes, make them less high to fit more width

        fireworkFlares.forEach((flare, i) => {
            const particle = this.scene.add.particles(`flare${i}`);
            particle.setDepth(depthDictionary.flares);
            this.particles.push(particle);
        });
    }

    renderBackground(
        windowWidth: number,
        windowHeight: number,
        trackLength: number,
        index: number,
        laneHeight: number,
        posY: number
    ) {
        let i = 0;
        const moon = this.scene.add.image(0, posY, 'moon');
        const newWidth = this.calcWidthKeepAspectRatio(moon, laneHeight);
        moon.setDisplaySize(newWidth, laneHeight);
        moon.setOrigin(0, 1);
        moon.setScrollFactor(0.1);
        moon.setDepth(depthDictionary.moon);
        this.backgroundElements?.push(moon);

        const repeats = Math.ceil(trackLength / newWidth) + 1;
        const backgroundKeys = ['starsAndSky', 'mountains', 'hills', 'trees', 'floor', 'grass'];
        const scrollFactors = [0.1, 0.4, 0.6, 0.9, 1, 1];
        const depths = [
            depthDictionary.sky,
            depthDictionary.mountains,
            depthDictionary.hills,
            depthDictionary.trees,
            depthDictionary.floor,
            depthDictionary.grass,
        ];

        while (i < repeats) {
            backgroundKeys.forEach((backgroundKey, idx) => {
                const element = this.scene.add.image(i * (windowWidth / this.numberPlayers), posY, backgroundKey);
                element.setDisplaySize(newWidth, laneHeight);
                element.setOrigin(0, 1);
                element.setScrollFactor(scrollFactors[idx]);
                element.setDepth(depths[idx]);

                // set new positions, based on size of image
                element.x = i * element.displayWidth;
                if (this.numberPlayers <= 2) element.y = this.moveLanesToCenter(windowHeight, laneHeight, index);
                this.backgroundElements?.push(element);
            });

            i++;
        }
    }

    private calcWidthKeepAspectRatio(bg: Phaser.GameObjects.Image | Phaser.GameObjects.TileSprite, laneHeight: number) {
        //Keep aspect ratio when setting displaySize
        // oldHeight .... 100%
        // newHeight ... x -> 100 / oldHeight * newHeight
        // ---
        // 100% ...... oldWidth
        // x ......... newWidth -> oldWidth/100 * x

        const oldHeight = bg.displayHeight;
        const backgroundScalingFactor = (100 / oldHeight) * laneHeight;
        const newWidth = (bg.displayWidth / 100) * backgroundScalingFactor;
        return newWidth;
    }

    private moveLanesToCenter(windowHeight: number, newHeight: number, index: number) {
        return (windowHeight - newHeight * this.numberPlayers) / 2 + newHeight * (index + 1);
    }

    renderPlayer(idx: number, coordinates: Coordinates, character: Character, username?: string): void {
        let usernameToDisplay = '';
        if (username) {
            usernameToDisplay = username;
        }

        if (!this.player) {
            this.renderPlayerInitially(coordinates, character.name);

            character.animations.forEach((animation: CharacterAnimation) => {
                this.initiateAnimation(character.name, animation.name, animation.frames);
            });
            this.renderPlayerName(idx, usernameToDisplay, coordinates.y);
        } else if (this.player) {
            //only move player
            this.player.x = coordinates.x;
        }
    }

    private renderPlayerName(idx: number, name: string, posY: number) {
        this.playerNameBg = this.scene.add.rectangle(50, posY - 25, 250, 50, 0xb63bd4, 0.7);

        this.playerName = this.scene.make.text({
            x: 20,
            y: posY - 30,
            text: `${name}`,
            style: {
                ...sharedTextStyleProperties,
                fontSize: `${16}px`,
                color: colors.white,
            },
            add: true,
        });

        this.playerNameBg.setDepth(depthDictionary.nameTag);
        this.playerName.setDepth(depthDictionary.nameTag);
        this.playerNameBg.setScrollFactor(0);
        this.playerName.setScrollFactor(0);
    }

    public updatePlayerNamePosition(newX: number, trackLength: number) {
        if (this.playerNameBg && this.playerNameBg.x < trackLength - this.scene.windowWidth) {
            this.playerNameBg?.setPosition(newX, this.playerNameBg.y);
            this.playerName?.setPosition(newX, this.playerName.y);
        }
    }

    renderChasers(chasersPositionX: number, chasersPositionY: number) {
        if (!this.chaser) {
            this.scene.anims.create({
                key: 'chasersAnimation',
                frames: this.scene.anims.generateFrameNumbers('chasersSpritesheet', { start: 0, end: 5 }),
                frameRate: 15,
                repeat: -1,
            });
            this.chaser = this.scene.physics.add.sprite(-1, chasersPositionY, 'chasersSpritesheet');
            this.chaser.setScale((0.7 / this.numberPlayers) * this.laneHeightsPerNumberPlayers[this.numberPlayers - 1]);
            this.chaser.setDepth(depthDictionary.chaser);
            this.chaser.y = this.chaser.y - this.chaser.displayHeight / 2; //set correct y pos according to player height
            this.chaser.play('chasersAnimation');
        }
        this.chaser.setX(chasersPositionX - 50); // - 50 so that not quite on top of player when caught
    }

    destroyPlayer() {
        this.player?.destroy();
    }

    stunPlayer(animationName: string) {
        const pebble = this.scene.physics.add.sprite(
            this.player!.x,
            this.backgroundElements![0].y - this.backgroundElements![0].displayHeight,
            'pebble'
        );
        pebble.setScale((0.4 / this.numberPlayers) * this.laneHeightsPerNumberPlayers[this.numberPlayers - 1]);
        pebble.y += pebble.displayHeight / 2;
        pebble.body.setGravity(0, 1200);
        pebble.setCollideWorldBounds(true);
        const destroyPebbleInterval = setInterval(() => {
            if (pebble.y > this.player!.y - (this.player!.displayHeight / 5) * 2.5) {
                clearInterval(destroyPebbleInterval);
                pebble.destroy();
                this.startAnimation(animationName);
            }
        }, 100);
    }

    renderCave(posX: number, posY: number) {
        posX -= 30; // move the cave slightly to the left, so the monster runs fully into the cave
        const scale = (0.9 / this.numberPlayers) * this.laneHeightsPerNumberPlayers[this.numberPlayers - 1];
        const yOffset = 2.2;
        this.caveBehind = this.scene.physics.add.sprite(posX, posY, 'caveBehind'); //TODO change caveBehind to enum
        this.caveBehind.setScale(scale);
        this.caveBehind.setDepth(depthDictionary.caveBehind);
        this.caveBehind.y -= this.caveBehind.displayHeight / yOffset; /// (0.01 * numberPlayers);

        this.caveInFront = this.scene.physics.add.sprite(posX, posY, 'caveInFront'); //TODO change caveInFront to enum
        this.caveInFront.setScale(scale);
        this.caveInFront.setDepth(depthDictionary.caveInFront);
        this.caveInFront.y -= this.caveInFront.displayHeight / yOffset; //(0.01 * numberPlayers);
    }

    handlePlayerDead() {
        //TODO change later - no need to color images that have already gone past
        if (this.backgroundElements && this.backgroundElements.length > 0) {
            this.backgroundElements.forEach(img => {
                img.setTint(0x123a3a);
            });
            const yPos = this.backgroundElements[0].y;
            const height = this.backgroundElements[0].displayHeight;
            const fixedWidth = 1200;

            const text = this.scene.make.text({
                x: this.scene.windowWidth / 2 - fixedWidth / 2,
                y: yPos - height / 2,
                text: 'The mosquito caught you. Look at your phone!',
                style: {
                    ...sharedTextStyleProperties,
                    fontSize: `${35}px`,
                    color: colors.orange,
                    stroke: colors.orange,
                    strokeThickness: 1,
                    fixedWidth,
                    align: 'center',
                    shadow: {
                        offsetX: 10,
                        offsetY: 10,
                        color: colors.black,
                        blur: 0,
                        stroke: false,
                        fill: false,
                    },
                },
                add: true,
            });
            text.scrollFactorX = 0;
            text.setDepth(depthDictionary.deadText);
        }
    }

    renderObstacles(posX: number, posY: number, obstacleScale: number, obstacleType: string, depth: number) {
        const obstacle = this.scene.physics.add.sprite(posX, posY, obstacleType);
        obstacle.setScale(obstacleScale * this.laneHeightsPerNumberPlayers[this.numberPlayers - 1]);

        obstacle.y -= obstacle.displayHeight / 1.3;
        obstacle.setDepth(depth);

        this.obstacles.push({ phaserInstance: obstacle });
    }

    renderFireworks(posX: number, posY: number, laneHeight: number) {
        const scales: Array<number | { min: number; max: number }> = [0.1, 0.08, { min: 0, max: 0.1 }];
        const lifespans: number[] = [250, 500, 700];

        this.particles.forEach((particle, i) => {
            const particlesEmitter = particle.createEmitter({
                // key: flare,
                x: posX,
                y: posY - (laneHeight / 4) * 3,
                scale: scales[i],
                speed: 75,
                blendMode: 'ADD',
                lifespan: lifespans[i],
            });
            particlesEmitter.on = true;
            setTimeout(() => (particlesEmitter.on = false), 900);
        });
    }

    movePlayerForward(newXPosition: number) {
        if (this.player) {
            this.player.x = newXPosition;
        }
    }

    destroyObstacle() {
        const currentObstacle = this.obstacles.shift();
        if (currentObstacle) {
            //     if (!currentObstacle.solvable) {
            //         currentObstacle.phaserInstance.destroy();
            //     } else {
            //         this.solvableObstacles.push(currentObstacle);
            //     }
            currentObstacle.phaserInstance.destroy();
        }
    }

    handleSkippedObstacle() {
        const currentObstacle = this.obstacles.shift();
        if (currentObstacle) this.skippedObstacles.push(currentObstacle);
    }

    destroyObstacles() {
        this.obstacles.forEach(obstacle => {
            obstacle.phaserInstance.destroy();
        });

        this.skippedObstacles.forEach(obstacle => {
            obstacle.phaserInstance.destroy();
        });
    }

    destroyCave() {
        this.caveBehind?.destroy();
        this.caveInFront?.destroy();
    }

    destroyChaser() {
        this.chaser?.destroy();
    }

    renderAttentionIcon() {
        if (!this.playerAttention && this.player) {
            this.playerAttention = this.scene.physics.add
                .sprite(
                    this.player.x + this.player.displayWidth / 2,
                    this.player.y - this.player.displayHeight / 2,
                    'attention'
                )
                .setDepth(depthDictionary.attention)
                .setScale((1 / this.numberPlayers) * this.laneHeightsPerNumberPlayers[this.numberPlayers - 1]);
        }
    }

    renderWarningIcon() {
        if (!this.playerWarning && this.player) {
            this.playerWarning = this.scene.physics.add
                .sprite(
                    this.player.x + this.player.displayWidth / 2,
                    this.player.y - this.player.displayHeight / 2,
                    'warning'
                )
                .setDepth(depthDictionary.attention)
                .setScale((1 / this.numberPlayers) * this.laneHeightsPerNumberPlayers[this.numberPlayers - 1]);
        }
    }

    destroyAttentionIcon() {
        this.playerAttention?.destroy();
        this.playerAttention = undefined;
    }

    destroyWarningIcon() {
        this.playerWarning?.destroy();
        this.playerWarning = undefined;
    }

    private renderPlayerInitially(coordinates: Coordinates, monsterSpriteSheetName: string) {
        this.player = this.scene.physics.add.sprite(coordinates.x, coordinates.y, monsterSpriteSheetName, 20);
        this.player.setDepth(depthDictionary.player);
        this.player.setBounce(0.2);
        this.player.setCollideWorldBounds(true);
        this.player.setScale((0.66 / this.numberPlayers) * this.laneHeightsPerNumberPlayers[this.numberPlayers - 1]);
        this.player.y = this.player.y - this.player.displayHeight / 2; //set correct y pos according to player height
    }

    private initiateAnimation(spritesheetName: string, animationName: string, frames: CharacterAnimationFrames) {
        this.scene.anims.create({
            key: animationName,
            frames: this.scene.anims.generateFrameNumbers(spritesheetName, frames),
            frameRate: 6,
            repeat: -1,
        });
    }

    startAnimation(animationName: string) {
        this.player?.play(animationName);
    }
    stopAnimation() {
        this.player?.anims.stop();
    }

    renderWind() {
        if (this.chaser) {
            this.scene.anims.create({
                key: 'windAnimation',
                frames: this.scene.anims.generateFrameNumbers('windSpritesheet', { start: 0, end: 5 }),
                frameRate: 8,
                repeat: 0,
            });
            this.wind = this.scene.physics.add.sprite(this.chaser.x - 50, this.chaser.y + 30, 'windSpritesheet');
            this.wind.setScale((0.5 / this.numberPlayers) * this.laneHeightsPerNumberPlayers[this.numberPlayers - 1]);
            this.wind.setDepth(depthDictionary.chaser);
            this.wind.y = this.wind.y - this.wind.displayHeight / 2; //set correct y pos according to player height
            this.wind.play('windAnimation');
            this.wind.on('animationcomplete', this.wind.destroy);
        }
    }
}
