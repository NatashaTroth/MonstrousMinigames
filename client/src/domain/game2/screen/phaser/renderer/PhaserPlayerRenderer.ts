import Phaser from 'phaser';

import { depthDictionary } from '../../../../../config/depthDictionary';
import SheepGameScene from '../../components/SheepGameScene';
import { Character, CharacterAnimationFrames } from '../gameInterfaces/Character';
import { Coordinates } from '../gameTypes/Coordinates';

/**
 * this is an incomplete PlayerRenderer adapter which contains all the phaser logic. This class might only be tested via
 * integration tests. That's why we want to keep this class as small as possible.
 */

export class PhaserPlayerRenderer {
    private player?: Phaser.Types.Physics.Arcade.SpriteWithDynamicBody;

    constructor(private scene: SheepGameScene) {}

    renderBackground() {
        const bg = this.scene.add.image(0, 0, 'forest2'); //TODO: replace bg image
        bg.setScale(0.4);
    }

    renderPlayer(coordinates: Coordinates, character: Character): void {
        if (!this.player) {
            this.renderPlayerInitially(coordinates, character.name);
            // frames:
            // 0-3: forward
            // 4-7: left
            // 8-11: right
            // TODO: missing walking back animation
            this.initiateAnimation(character.name, character.name + +'_walkForward', { start: 0, end: 3 });
            this.initiateAnimation(character.name, character.name + +'_walkLeft', { start: 4, end: 7 });
            this.initiateAnimation(character.name, character.name + +'_walkRight', { start: 8, end: 11 });
        } else if (this.player) {
            this.player.x = coordinates.x;
            this.player.y = coordinates.y;
        }
    }

    destroyPlayer() {
        this.player?.destroy();
    }

    movePlayerForward(newXPosition: number) {
        if (this.player) {
            this.player.x = newXPosition;
        }
    }

    renderSheepBackground(width: number, height: number) {
        const element = this.scene.add.image(0, 0, 'forest2');
        element.setDisplaySize(width, height);
        element.setOrigin(0, 1);
    }

    private renderPlayerInitially(coordinates: Coordinates, monsterSpriteSheetName: string) {
        this.player = this.scene.physics.add.sprite(coordinates.x, coordinates.y, monsterSpriteSheetName, 20);
        this.player.setScale(0.2);
        this.player.setDepth(depthDictionary.player);
        this.player.setCollideWorldBounds(true);
    }

    private initiateAnimation(spritesheetName: string, animationName: string, frames: CharacterAnimationFrames) {
        this.scene.anims.create({
            key: animationName,
            frames: this.scene.anims.generateFrameNumbers(spritesheetName, frames),
            frameRate: 6,
            repeat: -1,
        });
    }

    startAnimation(animationName: string) {
        this.player?.play(animationName);
    }
    stopAnimation() {
        this.player?.anims.stop();
    }

    movePlayer(character: Character, plusX?: number, plusY?: number) {
        if (plusY) {
            if (plusY > 0) {
                this.player?.play(`${character.name}_walkForward`);
            }
            if (plusY < 0) {
                this.player?.play(`${character.name}_walkBack`);
            }
            this.player?.setY(this.player?.y + plusY);
        }
        if (plusX) {
            if (plusX > 0) {
                this.player?.play(`${character.name}_walkRight`);
            }
            if (plusX < 0) {
                this.player?.play(`${character.name}_walkLeft`);
            }
            this.player?.setX(this.player?.x + plusX);
        }
    }
}
