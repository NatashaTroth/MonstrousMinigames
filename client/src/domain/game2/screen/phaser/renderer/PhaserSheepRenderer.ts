import Phaser from 'phaser';

import { depthDictionary } from '../../../../../config/depthDictionary';
import SheepGameScene from '../../components/SheepGameScene';
import { CharacterAnimationFrames } from '../gameInterfaces/Character';
import { Coordinates } from '../gameTypes/Coordinates';
import { SheepState } from '../Sheep';

/**
 * this is an incomplete PlayerRenderer adapter which contains all the phaser logic. This class might only be tested via
 * integration tests. That's why we want to keep this class as small as possible.
 */

export class PhaserSheepRenderer {
    private sheep?: Phaser.Types.Physics.Arcade.SpriteWithDynamicBody;

    constructor(private scene: SheepGameScene) {}

    preload() {
        // 0 - 4: right
        // 5 - 9: left
        // 10 - 13: forward
        // 14 - 17: back
        this.initiateAnimation('sheepSpritesheet', 'sheep_walkRight', { start: 0, end: 4 });
        this.initiateAnimation('sheepSpritesheet', 'sheep_walkLeft', { start: 5, end: 9 });
        this.initiateAnimation('sheepSpritesheet', 'sheep_walkForward', { start: 10, end: 13 });
        this.initiateAnimation('sheepSpritesheet', 'sheep_walkBack', { start: 14, end: 17 });
    }

    renderSheep(coordinates: Coordinates, sheepState: SheepState) {
        if (!this.sheep) {
            if (sheepState == SheepState.ALIVE) {
                this.renderSheepInitially(coordinates);
            } else if (sheepState == SheepState.DECOY) {
                this.placeDecoy();
            }
        } else if (this.sheep) {
            this.sheep.x = coordinates.x;
            this.sheep.y = coordinates.y;
        }
    }

    moveSheep(plusX?: number, plusY?: number) {
        if (plusY) {
            if (plusY > 0) {
                this.sheep?.play('sheep_walkForward');
            }
            if (plusY < 0) {
                this.sheep?.play('sheep_walkBack');
            }
            this.sheep?.setY(this.sheep?.y + plusY);
        }
        if (plusX) {
            if (plusX > 0) {
                this.sheep?.play('sheep_walkRight');
            }
            if (plusX < 0) {
                this.sheep?.play('sheep_walkLeft');
            }
            this.sheep?.setX(this.sheep?.x + plusX);
        }
    }

    destroySheep() {
        this.sheep?.destroy();
    }

    placeDecoy() {
        this.sheep?.setTexture('sheepDecoy');
    }

    private renderSheepInitially(coordinates: Coordinates) {
        this.sheep = this.scene.physics.add.sprite(coordinates.x, coordinates.y, 'sheepSpritesheet', 11);
        this.sheep.setDepth(depthDictionary.sheep);
        this.sheep.setCollideWorldBounds(true);
    }

    private initiateAnimation(spritesheetName: string, animationName: string, frames: CharacterAnimationFrames) {
        this.scene.anims.create({
            key: animationName,
            frames: this.scene.anims.generateFrameNumbers(spritesheetName, frames),
            frameRate: 6,
            repeat: -1,
        });
    }

    startAnimation(animationName: string) {
        this.sheep?.play(animationName);
    }
    stopAnimation() {
        this.sheep?.anims.stop();
    }
}
