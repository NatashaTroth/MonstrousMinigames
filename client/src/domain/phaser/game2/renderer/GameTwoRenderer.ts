/* eslint-disable no-console */
import Phaser from 'phaser';

import { depthDictionary } from '../../../../config/depthDictionary';
import SheepGameScene from '../../../game2/screen/components/SheepGameScene';
import * as colors from '../../colors';
import { loadingTextStyleProperties } from '../../textStyleProperties';

export class GameTwoRenderer {
    brightnessOverlay?: Phaser.GameObjects.Rectangle;
    guessInstructionText?: Phaser.GameObjects.Text;
    roundText?: Phaser.GameObjects.Text;
    sheepCountText?: Phaser.GameObjects.Text;

    constructor(private scene: SheepGameScene) {}

    renderBrightnessOverlay(width: number, height: number) {
        this.brightnessOverlay = this.scene.add.rectangle(0, 0, width, height, 0x000000, 1);
        this.brightnessOverlay.setDepth(depthDictionary.brightnessOverlay);
        this.brightnessOverlay.setDisplaySize(width, height);
        this.brightnessOverlay.setOrigin(0, 0);
        this.brightnessOverlay.setAlpha(0);
    }

    updateBrightnessOverlay(brightness: number) {
        this.brightnessOverlay?.setAlpha(this.getNewBrightness(brightness));
    }

    getNewBrightness(brightness: number) {
        return 1 - brightness / 100;
    }

    renderRoundCount(round: number) {
        if (this.roundText) {
            this.roundText.setText(`Round ${round}`);
        } else {
            const screenCenterWidth = this.scene.cameras.main.worldView.x + this.scene.cameras.main.width / 2;
            const screenCenterHeight = this.scene.cameras.main.worldView.y + 50;
            this.roundText = this.scene.make.text({
                x: screenCenterWidth,
                y: screenCenterHeight,
                text: `Round ${round}`,
                style: {
                    ...loadingTextStyleProperties,
                    fontSize: `30px`,
                    color: colors.orange,
                    fontStyle: 'bold',
                },
            });
            this.roundText.setOrigin(0.5);
            this.roundText.setDepth(depthDictionary.percentText);
        }
    }

    renderGuessText(show: boolean) {
        if (show) {
            if (!this.guessInstructionText) {
                const screenCenterWidth = this.scene.cameras.main.worldView.x + this.scene.cameras.main.width / 2;
                const screenCenterHeight = this.scene.cameras.main.worldView.y + this.scene.cameras.main.height / 2;
                this.guessInstructionText = this.scene.make.text({
                    x: screenCenterWidth,
                    y: screenCenterHeight - 50,
                    text: 'How many sheep are on the meadow?\nEnter your guess on your device.',
                    style: {
                        ...loadingTextStyleProperties,
                        fontSize: `${40}px`,
                        color: colors.orange,
                        fontStyle: 'bold',
                    },
                });
                this.guessInstructionText.setOrigin(0.5);
                this.guessInstructionText.setDepth(depthDictionary.percentText);
            } else {
                this.guessInstructionText.setVisible(true);
            }
        } else {
            this.guessInstructionText?.setVisible(false);
        }
    }

    renderSheepBackground(x: number, y: number, width: number, height: number) {
        const element = this.scene.add.image(x, y, 'sheepBackground');
        // const element = this.scene.add.image(x, y, 'sheepGrass');
        element.setDisplaySize(width, height);
        element.setOrigin(0, 0);
        element.setDepth(depthDictionary.game2Bg);
    }

    renderInitialSheepCount(count: number) {
        const screenCenterWidth = this.scene.cameras.main.worldView.x + this.scene.cameras.main.width / 2;
        const screenCenterHeight = this.scene.cameras.main.worldView.y + this.scene.cameras.main.height / 2;
        this.sheepCountText = this.scene.make.text({
            x: screenCenterWidth,
            y: screenCenterHeight - 50,
            text: `${count} sheeps are on the meadow`,
            style: {
                ...loadingTextStyleProperties,
                fontSize: `${60}px`,
                color: colors.black,
                fontStyle: 'bold',
                backgroundColor: colors.orange,
            },
        });
        this.sheepCountText.setPadding(10);
        this.sheepCountText.setOrigin(0.5);
        this.sheepCountText.setDepth(depthDictionary.percentText);
    }

    destroyInitialSheepCount() {
        this.sheepCountText?.destroy();
    }
}
