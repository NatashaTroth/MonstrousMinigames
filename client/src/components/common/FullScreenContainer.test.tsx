import { cleanup } from '@testing-library/react'
import React from 'react'
import renderer from 'react-test-renderer'

import FullScreenContainer from './FullScreenContainer'

afterEach(cleanup)
describe('FullScreenContainer', () => {
    it('renders given children', () => {
        const FullScreenContainerComponent = renderer
            .create(
                <FullScreenContainer>
                    <div>Text</div>
                </FullScreenContainer>
            )
            .toJSON()
        expect(FullScreenContainerComponent).toMatchSnapshot()
    })
})
