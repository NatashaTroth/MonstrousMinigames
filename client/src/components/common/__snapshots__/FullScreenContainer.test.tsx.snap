// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FullScreenContainer renders given children 1`] = `
<div
  className="sc-bdfBwQ fpPOZl"
>
  <div>
    Text
  </div>
</div>
`;
